Make SwitchCell easier to use with assistive technologies

adincebic opened the pull request:
Current implementation of SwitchCell does not support easier navigation with VoiceOver or other assistive technologies due to the fact that all UI elements are exposed as accessibility children. When cells have a switch and it is the only action button, tapping on the whole cell using assistive technologies should toggle its value. More over, it should not be separate accessibility element.

## To test

1. Turn voiceover on
2. Go to app settings from the profile tab
3. Tap on Notifications
4. Use swipe left or right gestures with one finger and try toggling notifications on and off

## Current behavior:

Voiceover treats label of the cell and the switch as two elements requiring users to swipe twice before reaching the switch. This gets more problematic if there are a lot of switches like in General section of app settings.

## Behaviour after this fix:

Voiceover treats cell label and switch control as single element and double tapping on it toggles its value just like in Apple's apps.

adincebic committed:
Make SwitchCell easier to use with assistive technologies

adincebic commented on podcasts/SwitchCell.swift:
```diff
@@ -46,4 +46,29 @@ class SwitchCell: ThemeableCell {
+
+    override func layoutSubviews() {
+        super.layoutSubviews()
+        setupAccessibility()
```
I did try calling this method from "awakeFromNib" but it was too early as "cellSwitch" and "cellLabels" had values defined in xib file which is not desirable.

adincebic commented on podcasts/SwitchCell.swift:
```diff
@@ -46,4 +46,29 @@ class SwitchCell: ThemeableCell {
+    }
+
+    private func setupAccessibility() {
+        isAccessibilityElement = true
```
I tried setting isAccessibilityElement to true, but it didn't work as voiceover for some reason tries to read accessibility label from its parent view (this cell).

adincebic commented on podcasts/SwitchCell.swift:
```diff
@@ -46,4 +46,29 @@ class SwitchCell: ThemeableCell {
+        let labelText = cellLabel.text ?? ""
+        let secondaryLabelText = cellSecondaryLabel.text ?? ""
+        accessibilityLabel = "\(labelText), \(secondaryLabelText)"
+        accessibilityTraits = cellSwitch.accessibilityTraits
```
I tried manually setting .switch as accessibility trait, but such trait does not exist. Accessibility trait for switch is not exposed and this is the only way to match behaviour like in Apple's apps.

fsalata replied:
To behave like Apple apps you only have to add the switch to the `accessoryView` instead of `contentView` of the cell.
I can't test on a real device right now but you can see in the video that the cell is read entirely and preforming the action (double tap) activates/deactivates the switch. 
Removing the switch from interface builder, changing line 4 to `var cellSwitch = ThemeableSwitch(frame: .zero)` and adding it to cell's `accessoryView` ( `self.accessoryView = cellSwitch`) should fix the behavior.
Hope it helps.

https://user-images.githubusercontent.com/7861088/197657365-78441bb8-3385-4bb1-ac9d-7c565460ecad.mov

adincebic replied:
> To behave like Apple apps you only have to add the switch to the `accessoryView` instead of `contentView` of the cell.
> 
> I can't test on a real device right now but you can see in the video that the cell is read entirely and preforming the action (double tap) activates/deactivates the switch. 
> 
> Removing the switch from interface builder, changing line 4 to `var cellSwitch = ThemeableSwitch(frame: .zero)` and adding it to cell's `accessoryView` ( `self.accessoryView = cellSwitch`) should fix the behavior.
> 
> Hope it helps.
> 
> 
> 
> https://user-images.githubusercontent.com/7861088/197657365-78441bb8-3385-4bb1-ac9d-7c565460ecad.mov
> 
> 
> 
> 

Thanks for the comment. I am aware of accessory view and accessibility behaviour when switch is its subview. I wanted to make as little changes as possible.

fsalata replied:
Oh ok.
You can try to hide the switch from the voiceover and set a `accessibilityCustomActions` for the cell to change the switch value programmatically

adincebic commented:
@emilylaguna  could you please add Accessibility label to this pr?

emilylaguna commented on podcasts/SwitchCell.swift:
```diff
@@ -46,4 +46,29 @@ class SwitchCell: ThemeableCell {
+    }
+
+    override func accessibilityActivate() -> Bool {
+        cellSwitch.setOn(!cellSwitch.isOn, animated: true)
```
If the isLocked value is set to false then we should return false here to prevent the switch from being activated.

You can reproduce by doing the following:

1. Sign into an account without plus
2. Navigate to the Apple Watch or Files settings
3. Select a Plus only switch
4. Activate it

emilylaguna commented on podcasts/SwitchCell.swift:
```diff
@@ -46,4 +46,29 @@ class SwitchCell: ThemeableCell {
+        matchAccessibilityValueWithSwitchControl()
+    }
+
+    private func matchAccessibilityValueWithSwitchControl() {
```
I think we can make this method name a bit more generic. 

```suggestion
    private func updateAccessibilityValue() {
```

emilylaguna commented on podcasts/SwitchCell.swift:
```diff
@@ -46,4 +46,29 @@ class SwitchCell: ThemeableCell {
+        isAccessibilityElement = true
+        let labelText = cellLabel.text ?? ""
+        let secondaryLabelText = cellSecondaryLabel.text ?? ""
+        accessibilityLabel = "\(labelText), \(secondaryLabelText)"
```
The secondaryLabel isn't actually used anywhere for the switch cell so we don't need to pass this value in here. In the future we should remove the label from the Xib and class. 
```suggestion
        accessibilityLabel = cellLabel.text
```

emilylaguna reviewed (changes requested):
Hey @adincebic, thank you so much for your contribution 🙌 👏 .

I gave this a test and everything is working with the exception of it bypassing the isLocked state (See the comment below) 

However I agree with @fsalata below that if we're updating the accessibility for this view we should also update it to reuse the system accessibility features by moving the switch into `accessoryView`.

What do you think about making that change?

adincebic commented:
> Hey @adincebic, thank you so much for your contribution 🙌 👏 .
> 
> 
> 
> I gave this a test and everything is working with the exception of it bypassing the isLocked state (See the comment below) 
> 
> 
> 
> However I agree with @fsalata below that if we're updating the accessibility for this view we should also update it to reuse the system accessibility features by moving the switch into `accessoryView`.
> 
> 
> 
> What do you think about making that change?

Sure, I'll get to it as soon as possible.

adincebic committed:
Remove cellSwitch and add it as accessory view

adincebic commented:
@emilylaguna  @fsalata  

* I moved cellSwitch to accessoryView
* Took isLocked into account
* Removed secondaryLabel from xib and Swift file

Please review.

adincebic commented:
Hey @fsalata  @emilylaguna  can I ask you to review this code once more? :)

fsalata commented on podcasts/SwitchCell.swift:
```diff
@@ -1,14 +1,12 @@
-    @IBOutlet var cellSwitch: ThemeableSwitch!
+    let cellSwitch: ThemeableSwitch = {
+        let cellSwitch = ThemeableSwitch()
+        cellSwitch.isAccessibilityElement = false
```
I believe you don't need to set it to false, when you add it to `accessoryView` it should work as expected

adincebic replied:
Not realy, it gets exposed as accessibility element along side parent cell. I tested it manually and needed to set it to false.

emilylaguna reviewed (approved):
Nice work! This works nicely in my testing. 

Thank you so much for your contribution! I'll get this merged in.
